### PR TITLE
Added a rather elaborate example on how Layered Attestion could take place.

### PR DESCRIPTION
**Pull request by henkbirkholz**
Please feel free to bash. This is a basis for discussion.

### CODE REVIEW

File: draft-ietf-rats-architecture.md [hunk 1]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
...
 embedded Secure Elements (eSE), or Hardware Security Modules (HSM).
 
+## Layered Attestation Procedures
+
+By definition, the Attester role takes on the duty to create Attestation Evidence.
+The fact that an Attester role is composed of several types of environments that
+can be nested or staged adds complexity to the architectural layout of how an
+Attester -- in itself -- is composed and therefore has to conduct the Claims Collection
+in order to create believable Attestation Evidence. This complexity is
+independent of the concept defined by Composite Attesters.
```
Comment (mcr) [inline draft-ietf-rats-architecture.md]: Maybe the point at the end, about how a composite attester can each produce their own *signed* evidence could be brought to the top.

File: draft-ietf-rats-architecture.md [hunk 2]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
 Attesting Environments include Trusted Execution Environments (TEE),
 embedded Secure Elements (eSE), or Hardware Security Modules (HSM).
 
+## Layered Attestation Procedures
+
+By definition, the Attester role takes on the duty to create Attestation Evidence.
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: s/Attestation Evidence/Evidence/

File: draft-ietf-rats-architecture.md [hunk 3]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
...
+The fact that an Attester role is composed of several types of environments that
+can be nested or staged adds complexity to the architectural layout of how an
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: not sure what "staged" means

File: draft-ietf-rats-architecture.md [hunk 4]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
...
+Attester -- in itself -- is composed and therefore has to conduct the Claims Collection
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: nit: Collection should not be capitalized

File: draft-ietf-rats-architecture.md [hunk 5]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
...
+in order to create believable Attestation Evidence. This complexity is
+independent of the concept defined by Composite Attesters.
+Composite Attesters are inherently composed of multiple nested environments.
+The following example is intended to illustrate this composition:
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: wording: 388 and 390 seem redundant

File: draft-ietf-rats-architecture.md [hunk 6]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
...
+
+A very common example is elaborated on to illustrate Layered Attestation.
+
+The very first Attesting Environment has to ensure the integrity of
+the (U)EFI / BIOS / Firmware that initially boots up a composite device (e.g.
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: nit: always put a comma after e.g.

File: draft-ietf-rats-architecture.md [hunk 7]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
...
+a cell phone). A hardware component (which effectively can encompasses
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: nit: too many parenthetical statements makes this hard to read

File: draft-ietf-rats-architecture.md [hunk 8]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
...
+immutable execution environments and mutable execution environments)
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: is the meaning of "immutable execution environment" clear enough?   I'm assuming you mean that the code is immutable, but the data is not.

File: draft-ietf-rats-architecture.md [hunk 9]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
...
+executes the very first Claim Collection. These Claims MUST be stored
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: we don't use MUST etc in this doc, just plain english

File: draft-ietf-rats-architecture.md [hunk 10]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
...
+securely (e.g. a eSE. It is not ensured that a TEE or HWM are accessible
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: undefined acronyms, e.g., what's HWM?

File: draft-ietf-rats-architecture.md [hunk 11]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
...
+at this very beginning). At this stage of the boot-cycle of a
+thing, the Claims collected typically cannot be composed into Attestation Evidence.
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: s/thing/composite device/ ?

Comment (dthaler) [inline draft-ietf-rats-architecture.md]: This is hard to follow... why can't they composed into evidence?

File: draft-ietf-rats-architecture.md [hunk 12]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
...
+
+The very first Attesting Environment in this example can be a static root-of-trust
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: capitalization "Root of Trust" for consistency with rest of doc.  But what does "static" mean here?   Suggest deleting that word.

File: draft-ietf-rats-architecture.md [hunk 13]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
...
+provided by a hardware component residing next to the (U)EFI / BIOS / Firmware.
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: delete "residing next to the (U)EFI / BIOS / Firmware"

rationale: I don't what it means to have hardware "residing next to" code.

File: draft-ietf-rats-architecture.md [hunk 14]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
...
+As in any other scenario, this hardware component is is the first Attesting Environment.
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: s/is is/is/

File: draft-ietf-rats-architecture.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (dthaler) [reply draft-ietf-rats-architecture.md]: Agree with Michael.  Also, there can be multiple layers in there where UEFI is not the thing that the hardware first starts.  Example:
ROM -> SPL -> UEFI -> ...

File: draft-ietf-rats-architecture.md (hunk 5 above)
Comment (dthaler) [reply draft-ietf-rats-architecture.md]: Ned's diagram may fit here

File: draft-ietf-rats-architecture.md [hunk 15]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
...
+most important feature of layered attestation:
+The successfully measured [comment: this is why Henk prefers 'measured' above 'collected'
+because it can be used in English text better] environment that is the
+(U)EFI / BIOS / Firmware now becomes the Attesting Environment.
+
+Now, the duties have been transferred and Layered Attestation takes place.
+The initial Attesting Environment relinquishes its duties to the Target Environment.
+It is important to note that the new Attesting Environment cannot alter
+the content about its own measurements. If the Attesting Environment
+would be able to do that, Layered Attestation would become unfeasible.
```
[the remark this answers is not in the document]
Comment (dthaler) [reply draft-ietf-rats-architecture.md]: @mcr: yes the RoT has signed the measurement.   a TPM/TEE/etc is not required to do such signing.

File: draft-ietf-rats-architecture.md [hunk 16]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
...
+additional Claims and is storing them in a secure and shielded manner.
+
+The essence of this example is a cascade of staged boot environments. Each
+environment (after the initial one that is a root-of-trust) has the duty
+of measuring its next environment before it is started. Therefore, creating
+a layered boot sequence and correspondingly enabling Layered Attestation.
+
+Please note: nested components can compose their on Composite Attesters.
+In consequence, the concept of Layered Attestation is realized in each Attester role.
+If there is a Composite Device enabling Composite Attesters, each Attester in this composite
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: delete "in this composite"

File: draft-ietf-rats-architecture.md [hunk 17]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
...
+can enable its separate instance of Layered Attestation. Layered Attestation takes place
+depending on the complexity of the composition of each Attester role in a Composite Attester.
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: I can't understand the last sentence above.  Suggest deleting.

File: draft-ietf-rats-architecture.md [hunk 18]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
...
+In this example the duty of being the Attesting Environment is now
+taken over by the (U)EFI / BIOS / Firmware that was the Attested
+Environment before. This transfer of duty is the essential part of
+Layered Attestation. The (U)EFI / BIOS / Firmware now is the Attesting Environment.
+The next Target Environment is, in this example, a bootloader. There are
+potentially multiple kernels to boot, the decision is imposed on the bootloader.
+Only a integer bootloader will make an appropriate decision. Therefore, Claims about
+the integrity of a bootloader are now collected by the freshly appointed Attesting Environment
+that is the (U)EFI / BIOS / Firmware. Collected Claims have to be stored by the current
+Attesting Environment in a similar shielded and secured manner, so that the next Attesting Environment
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: No idea what "shielded" means in this context, but if the claims have been signed then why is that not sufficient?

File: draft-ietf-rats-architecture.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: This is example of "staged". Analogously, I think you are correct, "staged" is not well defined and has to be elaborated on.

File: draft-ietf-rats-architecture.md (hunk 8 above)
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: You are correct, this is not clear enough. Will be addressed.

File: draft-ietf-rats-architecture.md (hunk 9 above)
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: Force of habit... willl use "have to"

File: draft-ietf-rats-architecture.md (hunk 10 above)
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: I assumed we used that already. Did not check. It is a hardware security module, e.g. a S-CRTM (Static Code Root of
Trust for Measurement)

File: draft-ietf-rats-architecture.md (hunk 11 above)
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: more specific (apparently) but okay

File: draft-ietf-rats-architecture.md (hunk 18 above)
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: Because I am not sure that every Root of Trust for Measurement does actually sign the measurements. Is that the case? "Shielding" is derived from "shielded location" we should maybe use instead: "A place (memory, register, etc.) where it is safe to operate on sensitive data; data locations that can be accessed only by Protected Capabilities."). Shielded Locations are required to store information, if it is not signed. A can be TEE a shielded location.

File: draft-ietf-rats-architecture.md (hunk 17 above)
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: Okay

File: draft-ietf-rats-architecture.md [hunk 19]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
...
+the content about its own measurements. If the Attesting Environment
+would be able to do that, Layered Attestation would become unfeasible.
+
+In this example the duty of being the Attesting Environment is now
+taken over by the (U)EFI / BIOS / Firmware that was the Attested
+Environment before. This transfer of duty is the essential part of
+Layered Attestation. The (U)EFI / BIOS / Firmware now is the Attesting Environment.
+The next Target Environment is, in this example, a bootloader. There are
+potentially multiple kernels to boot, the decision is imposed on the bootloader.
+Only a integer bootloader will make an appropriate decision. Therefore, Claims about
```
[the remark this answers is not in the document]
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: German English... "a bootloader with intact integrity"

File: draft-ietf-rats-architecture.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: eSE is an ARM term, I think. "embedded secure element" It was often referred to with respect to the EAT I-D discussions. We can just say RTM (Root of Trust for Measurement). I am still trying not to include the term trust too often.

File: draft-ietf-rats-architecture.md (hunk 11 above)
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: Because there is no Root of Trust for Reporting or Attesting Environment  (e.g. a TEE with a TPM) available yet.

File: draft-ietf-rats-architecture.md (hunk 16 above)
Comment (nedmsmith) [inline draft-ietf-rats-architecture.md]: I'm confused by the conversation about Composite Device and the unknown thing called Composite Attester in the context of a description of a "Layered Device?" or "Layered Attestation". It seems to me the architecture should clearly describe each independently then let implementation dictate if they should be combined. Combining them shouldn't result in new architecture - not unless there is some anomaly lurking that warrants it.

File: draft-ietf-rats-architecture.md [hunk 20]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
 Attesting Environments include Trusted Execution Environments (TEE),
 embedded Secure Elements (eSE), or Hardware Security Modules (HSM).
 
+## Layered Attestation Procedures
```
Comment (nedmsmith) [inline draft-ietf-rats-architecture.md]: I think a picture is worth 1000 words. Dave T had an ascii art example that seemed pretty good. Is there a reason not to include it here?

File: draft-ietf-rats-architecture.md (hunk 5 above)
Comment (nedmsmith) [reply draft-ietf-rats-architecture.md]: Dave had the ascii art version. Generally speaking it is easier to explain a complex concept by providing a diagram that illustrates the concept. Details about the concept can be organized according to a decomposition of the diagram. If we do this first then examples can be provided for intuition and don't need excessive explanation. We could just say "device bootstrap sequence is an example of a 'layered device' where each phase of boot implements a RATS Attester that collects claims about the next phase".

Comment (nedmsmith) [reply draft-ietf-rats-architecture.md]: Maybe there are other use cases that provide intuition - maybe a hypervisor loading a guest VM?

File: draft-ietf-rats-architecture.md [hunk 21]
```diff
@@ -376,6 +376,76 @@ create Evidence. Examples of environments that can be used as
...
+securely (e.g. a eSE. It is not ensured that a TEE or HWM are accessible
+at this very beginning). At this stage of the boot-cycle of a
+thing, the Claims collected typically cannot be composed into Attestation Evidence.
+
+The very first Attesting Environment in this example can be a static root-of-trust
+provided by a hardware component residing next to the (U)EFI / BIOS / Firmware.
+As in any other scenario, this hardware component is is the first Attesting Environment.
+It collects a rather concise number of Claims about the Target Environment.
+The Target Environment in this example is the (U)EFI / BIOS / Firmware.
+After the boot sequence started the Target Environment is conducts the
```
[the remark this answers is not in the document]
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: This was not a sentence. I hope now it is. Added a clarifying sentence, too.

File: draft-ietf-rats-architecture.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: removed "(e.g. a eSE. It is not ensured that a TEE or HWM are accessible
at this very beginning)" in the paragraph before. Does this fix the arbitrary restriction created by too many names for RoT?

File: draft-ietf-rats-architecture.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: I have no good editorial solution (from myself or others). We have to discuss this, unfortunately.

File: draft-ietf-rats-architecture.md (hunk 1 above)
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: I am unsure, what the final decision on this is / will be. Will have to discuss

File: draft-ietf-rats-architecture.md (hunk 2 above)
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: Did "fix" that, although I do not see the difference? Therefore left open.

File: draft-ietf-rats-architecture.md (hunk 3 above)
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: Introduced the term Staging at the end of the 3rd paragraph. I think this is better than not using "staging", it is an intuitive term that does not has to be introduces, necessarily. Any other thoughts?

File: draft-ietf-rats-architecture.md (hunk 5 above)
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: Could we fix this together please? I find it difficult to make a editorial decision here without further input

File: draft-ietf-rats-architecture.md (hunk 7 above)
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: Most likely fixed.

File: draft-ietf-rats-architecture.md (hunk 8 above)
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: added mor clarifying parenthesises... not everybody is a fan of that. We should look at that.

File: draft-ietf-rats-architecture.md (hunk 10 above)
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: relates to the added clarification about ""immutable execution environment". What words are most appropriate here? Static Code Root of Trust for Measurement is very precise and fits the context. But these are also a lot of words and people seem to be easily confused by them.

File: draft-ietf-rats-architecture.md (hunk 12 above)
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: it means "static code", see above. Added code to be conistent

File: draft-ietf-rats-architecture.md (hunk 16 above)
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: fixed

File: draft-ietf-rats-architecture.md (hunk 18 above)
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: I assume we have to discuss this, unfortunately

File: draft-ietf-rats-architecture.md (hunk 16 above)
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: This is a weird artifact while we were trying out new terms in this context. Could we go through this text passage together, please? There is no single good answer it seems to me.

File: draft-ietf-rats-architecture.md (hunk 20 above)
Comment (henkbirkholz) [reply draft-ietf-rats-architecture.md]: Let's have look and maybe a PR.

File: draft-ietf-rats-architecture.md [hunk 22]
```diff
@@ -376,6 +376,73 @@ create Evidence. Examples of environments that can be used as
...
+independent of the concept defined by Composite Attesters.
+Composite Attesters are inherently composed of multiple nested environments.
+The following example is intended to illustrate this composition:
+
+A very common example is elaborated on to illustrate Layered Attestation.
+
+The very first Attesting Environment has to ensure the integrity of
+the (U)EFI / BIOS / Firmware that initially boots up a composite device (e.g.,
+a cell phone). A hardware component, which effectively can encompasses
+immutable execution environments (static code Roots of Trust) and mutable execution environments (dynamic code Roots of Trusts),
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: I don't understand what it means for a hardware component to "encompass .. mutable execution environments (dynamic roots of trust)".

File: draft-ietf-rats-architecture.md [hunk 23]
```diff
@@ -376,6 +376,73 @@ create Evidence. Examples of environments that can be used as
...
+executes the very first Claim Collection. These Claims have to be stored
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: The term "stored" sounds like persisted storage, which I don't think is required

File: draft-ietf-rats-architecture.md [hunk 24]
```diff
@@ -376,6 +376,73 @@ create Evidence. Examples of environments that can be used as
...
+composite device, the Claims collected typically cannot be composed into Evidence.
+
+The very first Attesting Environment in this example can be a hardware component that is a Static Code Root of Trust.
+As in any other scenario, this hardware component is the first Attesting Environment.
+It collects a rather concise number of Claims about the Target Environment.
+The Target Environment in this example is the (U)EFI / BIOS / Firmware.
+After the boot sequence started, the Target Environment conducts the
+most important and defining feature of layered attestation:
+The successfully measured [comment: this is why Henk prefers 'measured' above 'collected'
+because it can be used in English text better] environment that is the
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: should delete the [comment] before merging

File: draft-ietf-rats-architecture.md [hunk 25]
```diff
@@ -376,6 +376,73 @@ create Evidence. Examples of environments that can be used as
...
+(U)EFI / BIOS / Firmware now becomes the Attesting Environment. Analogously, the Attesting Environment hands off its duty to one of its Target Environments. This procedure in Layered Attestation is called Staging.
```
Comment (dthaler) [inline draft-ietf-rats-architecture.md]: I haven't heard the term Staging before, not sure why we need to mention it here
